File: python/Add_binary.py
'''
Created on 2014-3-23

Leetcoder : Add binary

Given two binary strings, return their sum (also a binary string).

For example,
a = "11"
b = "1"
Return "100".

@author: xqk
'''

class Solution:
    # @param a, a string
    # @param b, a string
    # @return a string
    def addBinary(self, a, b):
        if len(a) == 0:
            return b
        if len(b) == 0:
            return a;
        if len(a) < len(b):
            temp = a
            a = b
            b = temp
        r = [];
        f = 0
        ap = len(a) -1
        bp = len(b) - 1
        while bp>=0:
            sumV = int(b[bp])+int(a[ap])+f
            f = sumV/2
            
            r.append(str(sumV%2) );
            bp = bp - 1
            ap = ap -1 
        while ap >= 0:
            
            sumV = int(a[ap])+f
            f = sumV/2
           
            r.append(str(sumV%2) );
            ap = ap -1
        s = ''
        for i in range(len(r)):
            s= s + r[len(r) -1 -i]
        
        if f == 1:
            s = '1'+s
        return s
        
        
        
        
a = '100'
b='110010'
test = Solution()
print test.addBinary(a, b)

File: Cplusplus/Add_Binary.cpp
/*
Leetcode : Add_Binary.cpp

Given two binary strings, return their sum (also a binary string).

For example,
a = "11"
b = "1"
Return "100".
*/
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
            int carry = 0;
            string result = "";
            if(a.size() < b.size() )
            {
                string tmp = b;
                b = a;
                a = tmp;
            }
            int lB = b.size();
            int lA = a.size();
            int pos = 0;
            while(pos < lB && pos < lA)
            {
                int value = carry + int(b[lB - 1-pos]) - 48 + int(a[lA - 1 -pos]) - 48;
                pos++;
                switch (value)
                {
                    case 0:
                        result.append("0");
                        break;
                    case 1:
                        result.append("1");
                        carry = 0;
                        break;
                    case 2:
                        carry = 1;
                        result.append("0");
                        break;
                    case 3:
                        carry = 1;
                        result.append("1");
                        break;

                }
            }
            for(int i = pos ; i < lA ; i++)
            {
                if(carry)
                {
                    if(a[lA - 1- pos] == '1')
                    {
                        result.append("0");
                    }
                    else
                    {
                        result.append("1");
                        carry = 0;
                    }
                }
                else
                    result.push_back(a[lA - 1- i]);
            }
            if(carry)
                result.append("1");
            reverse(result.begin() , result.end());
            return result;
    }
};

int main()
{
    Solution test = Solution();
    string a = "11";
    string b = "1";
    cout<<test.addBinary(a,b)<<endl;
}
